_source/_posts/2021-08-13-reactive-java.md

```diff
@@ -0,0 +1,787 @@
+
+A probably simplified [marble diagam](https://projectreactor.io/docs/core/release/reference/#howtoReadMarbles) for the example above might look like this:
+
+{% img blog/reactive-java/marble-diagram.png alt:"Marble Diagram" width:"800" %}{: .center-image }
+
+## Reactive Java Spring Services
+
+In Spring WebFlux, it is assumed applications don't block, so non-blocking servers use a small fixed-size thread pool to handle requests, named **event loop** workers.
+
+In an ideal reactive scenario, all the architecture components are non-blocking, so there is no need to worry about the event loop freezing up (**reactor meltdown**). But sometimes, you will have to deal with legacy blocking code, or blocking libraries.
+
+So now, let's experiment with Reactor Schedulers in a Reactive Java application. Create a Spring WebFlux service with Okta security. The service will expose an endpoint to return a random integer.
+
+The implementation will call Java `SecureRandom` blocking code. Start by downloading a Spring Boot Maven project using [Spring Initializr](https://start.spring.io/). You can do it with the following HTTPie command:
+
+```shell
+http -d https://start.spring.io/starter.zip \
+  bootVersion==2.5.3 \
+  baseDir==reactive-service \
+  groupId==com.okta.developer.reactive \
+  artifactId==reactive-service \
+  name==reactive-service \
+  packageName==com.okta.developer.reactive \
+  javaVersion==11 \
+  dependencies==webflux,okta
+```
+
+Unzip the project:
+
+```shell
+unzip reactive-service.zip
+cd reactive-service
+```
+
+{% include setup/cli.md type="web" framework="Okta Spring Boot Starter" %}
+
+Create the package `com.okta.developer.reactive.service`, and add the `SecureRandomService` interface:
+
+```java
+package com.okta.developer.reactive.service;
+
+import reactor.core.publisher.Mono;
+
+public interface SecureRandomService {
+
+    Mono<Integer> getRandomInt();
+}
+```
+
+And the implementation `SecureRandomServiceImpl`:
+
+```java
+package com.okta.developer.reactive.service;
+
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Mono;
+
+import java.security.SecureRandom;
+
+@Service
+public class SecureRandomServiceImpl implements SecureRandomService {
+
+    private SecureRandom secureRandom;
+
+    public SecureRandomServiceImpl() {
+        secureRandom = new SecureRandom();
+    }
+
+    @Override
+    public Mono<Integer> getRandomInt() {
+        return Mono.just(secureRandom.nextInt());
+    }
+}
+```
+
+Think for a moment about the `getRandomInt` implementation. It's hard to resist the temptation to wrap anything whatsoever in a publisher. `Mono.just(T data)` creates a `Mono` that will emit the specified item. But the item is **captured at instantiation time**. This will invoke the blocking code at assembly time, perhaps in the context of an event loop thread.
```

Comment:
```suggestion
Think for a moment about the `getRandomInt()` implementation. It's hard to resist the temptation to wrap anything whatsoever in a publisher. `Mono.just(T data)` creates a `Mono` that will emit the specified item. But the item is **captured at instantiation time**. This will invoke the blocking code at assembly time, perhaps in the context of an event loop thread.
```